ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/P2TargetUtils.java

```diff
@@ -1547,6 +1551,24 @@ private URI[] getMetadataRepositories(ITargetDefinition target) throws CoreExcep
 		return result.toArray(new URI[result.size()]);
 	}
 
+	private List<IInstallableUnit> getAdditionalProvisionIUs(ITargetDefinition target) {
+		List<IInstallableUnit> result = new ArrayList<>();
+		ITargetLocation[] containers = target.getTargetLocations();
+		if (containers != null) {
+			for (ITargetLocation container : containers) {
+				if (container instanceof IUBundleContainer) {
+					// this is already handled by getMetadataRepositories(..)
+					continue;
+				}
+				IInstallableUnitQueryable queryable = Adapters.adapt(container, IInstallableUnitQueryable.class);
```

Comment:
Instead of Using `Adapters` and requiring a special sub-class of `IQueryable`, can't we use the existing `ITargetLocation.getBundles()/getFeatures()` and wrap the returned `TargetBundle/TargetFeature` instance to an `IInstallableUnit`?
Maybe this would require to change the order in `TargetDefinition.resolve()` in which currently IU-Locations are resolved first.
This would also have the advantage that there are no adjustments in other Locatation-Types (like Maven Locations) necessary.

Reply:
> Instead of Using Adapters and requiring a special sub-class of IQueryable, can't we use the existing ITargetLocation.getBundles()/getFeatures() and wrap the returned TargetBundle/TargetFeature instance to an IInstallableUnit?

Not really as there is no guarantee that the set of *resolved* bundles an features is the full set of available items.
Just assume the IU case:

- I have an update site that contains 100 IUs
- My location selects only one that has no dependencies so `getBundle()` will return one IU
- No there is another location that has a requirement to one of the not selected items
- --> resolve error as the other 99 IU are not visible through this approach

Beside that there would be a risk of endless recursion if `A` want the bundles of `B` and B then in return queries `A`.

> Maybe this would require to change the order in TargetDefinition.resolve() in which currently IU-Locations are resolved first.

This will require defiantly changes there, I'm actually not very happy with all this specific handling of concrete types but that's another story.

> This would also have the advantage that there are no adjustments in other Locatation-Types (like Maven Locations) necessary.

I don't think its possible to solve this without changing the implementation of location types as there currently is no way to know what else *might* be available.

Reply:
> > This would also have the advantage that there are no adjustments in other Locatation-Types (like Maven Locations) necessary.
> 
> I don't think its possible to solve this without changing the implementation of location types as there currently is no way to know what else _might_ be available.

Agree on that, but nevertheless I think this could all be solved in PDE.
As far as I can overview it, what's effectively done, is to collect all target-bundles and -features of all location and wrap them into IUs.
But I don't see a need why this has to be solved via the Adapter-Approach (in general I find that concept bad because it effectively defines an API but does not look like one and is hard to find).
Why can't we just add an Interface like `IIUProviderTargetLocation` (or whatever is suitable) to **PDE** that has a method `getAllBundles()` and `getAllFeature()`? Then this could be called here without the indirection via the Adapter and an `IQueryable`?
Sure, we probably have to define another interface but I think it should be defined in PDE instead of P2, because from what I see now it is just used for target-locations and has no value for P2.

This would also have the advantage that for example in M2E or Oomph the internal `InstallableUnitGenerator` has not to be referenced or reimplemented to provide the IUs.

Thus I think https://github.com/eclipse-equinox/p2/pull/348 is not necessary and should be reverted to not unnecessarily add extra interfaces to P2 that only serve PDE and make the integration more complex.

Or is there something in the bigger picture that I fail to see?

Reply:
Adapter pattern is widely used in Eclipse and PDE and makes the code more flexible (e.g. you don't need to actually implement that in the class itself but can have a separate wrapper/adapter) and less coupled so I don't se a reason not using it here. Especially as I don't want to introduce an "API" this is completely internal behavior of the `IUBundleLocation` and not of general interest for others.

That's also why I don't want something like a `IIUProviderTargetLocation` or similar as this will only manifest the already bad design of existing Target Location API design that requires some materialized bundles / features while in fact we want to operate on the **metadata** of the IUs. That's also a reason why PDE is ultra-slow in resolving such locations compared to Tycho.

> This would also have the advantage that for example in M2E or Oomph the internal InstallableUnitGenerator has not to be referenced or reimplemented to provide the IUs.

P2 and PDE actually require to reference "internal" stuff all the way and m2e was always requiring latest PDE to support best integration so I don't really  mind that. For Targelts I also can thing that Oomph already operates on P2 InstallableUnits and don't need a converter anyway.

Most important, these changes are more intermediate solutions to lower the pain we currently see, my main goal is to rework how PDE TargetLocations are implemented in general to using the OSGi Repository API that will be much more flexible and complete than anything we can archive with the current stuff where we must respect so much legacy..

Reply:
> Especially as I don't want to introduce an "API" this is completely internal behavior of the `IUBundleLocation` and not of general interest for others.

But effectively you make this an API here, because if you remove the code something does not work as before. Behavior is also API although the API-Tools or similar checks don't cover that. That's why I don't like the Adapter approach. It often does not feel like an API but effectively it is. Any way.


> Most important, these changes are more intermediate solutions to lower the pain we currently see, my main goal is to rework how PDE TargetLocations are implemented in general to using the OSGi Repository API that will be much more flexible and complete than anything we can archive with the current stuff where we must respect so much legacy..

Acknowledged that and that sounds like a great plan. But from my experience resolving does not download artifacts, but it still is not that fast. :) I assume it works on the Plan the Planner creates to update the Target's profile.

Back to topic, I'm still convinced we can just use `IQueryable<IInstallableUnit>` here.
The only drawback is that `ExpressionMatchQuery` does not perform an actual type-check if the operand is `ExpressionUtil.TRUE_EXPRESSION`. Therefore an explicit type-check is necessary to filter IUs and we cannot rely on the generic type. But thanks to the new `IQueryResult.stream()` method this is quite simple. :)

I'm about to push another commit on top of your branch to demonstrate it.
With that I would be more than happy to have this great new feature. :)

Reply:
> But from my experience resolving does not download artifacts

"Resolving" a target location requires to download every item to disk because the `TargetBundle` requires a physical file,for `TargetFeature` I have already lifted that requirement to support virtual features but if  a feature is not virtual it again requires a physical download as you can't recover the feature model from the meta-data in almost all cases.